Repackage the Alembic migration test as a single tox-based script

webbnh opened the pull request:
This PR builds on #3201 to repackage the Alembic migration test so that it can be invoked directly via `tox` after the linters which are run in `build.sh`.

Changes:
- `jenkins/run-alembic-migrations-check`:
  - delete this file, moving its contents to either `build.sh` or `lib/pbench/server/database/alembic.check`
- `build.sh`:
  - invoke `tox` to run the Alembic migration check after running the linters
- `lib/pbench/server/database/alembic.check`:
  - this script now performs the whole test, including starting and stopping the PostgreSQL container
  - made some small changes to replace relative file paths with absolute ones, since the script changes the default directory in the middle
- `tox.ini`:
  - added two `jenkins/run`-related environment variables to the set passed into the Tox environment
- `jenkins/Pipeline.gy`:
  - remove the explicit pipeline stage for the migration check, since the check is now run as part of the subsequent stage

Suggested commit log:
```
Repackage the Alembic migration test

- run it under build.sh instead of as a separate pipeline stage
- create the PostgreSQL container from inside the test script
- delete the wrapper script

PBENCH-1060
```

dbutenhof commented on jenkins/Pipeline.gy:
```diff
@@ -2,7 +2,7 @@ pipeline {
     environment {
         // Environment variables for various `jenkins/run` invocations.
-        EXTRA_PODMAN_SWITCHES="--pull=always -e=COV_REPORT_XML=yes"
+        EXTRA_PODMAN_SWITCHES="--pull=always -e=COV_REPORT_XML=yes --network host"
```
Not necessarily objecting to making this the default (I think we probably either need it everywhere, or at least it wouldn't hurt); but why add here *and* in `build.sh`? Yes, we might run `build.sh` "offline", but then do we really need to change the default here?

webbnh replied:
The reason for doing it in `build.sh` is that the invocation of the `alembic-check` needs to have `--network host` (while most invocations of `jenkins/run` actually don't, although AFAIK it doesn't hurt to have it for them), so, in order to make `build.sh` useful _outside_ the CI environment, we need to specify it in the invocation in there.  Moreover, the invocation of `alembic-check` does _not_ need the `COV_REPORT_XML` environment variable (and it's probably better off without it), so overriding the default value of `EXTRA_PODMAN_SWITCHES` in `build.sh` is arguably a plus.

That said, I guess we don't need to add `--network host` to the pipeline default, since I'm not aware of any other uses of `EXTRA_PODMAN_SWITCHES` which need the switch.  However, as I said, I don't think adding it hurts anything, and having it there will cover any future cases of cooperating container invocations.

One note is that the redefinition of `EXTRA_PODMAN_SWITCHES` in `build.sh` overrides any pull policy set in the outer environment (like the pipeline default).  In the CI, this shouldn't have any effect, because `build.sh` is invoked via `jenkins/run` in a pipeline stage, and so _that_ invocation will have pulled a fresh copy of the container, per the pipeline default `EXTRA_PODMAN_SWITCHES` value, so _not_ having `--pull always` defined for the inner invocation is actually a plus there, as well, since it avoids a pointless pull when running under the CI.  (For use of `build.sh` outside the CI, whether/how `EXTRA_PODMAN_SWITCHES` is defined in the developer's environment (and whether a fresh container image needs to be pulled) is up to the developer.)

webbnh replied:
> as I said, I don't think adding it hurts anything

And, apparently, I was wrong!  Adding the `--network host` causes `/sys/kernel` inside the container to be overmapped with an empty `tmpfs` file system, which breaks the Agent legacy `datalog` `sysfs` test by preventing it from finding the `/sys/kernel/vmcoreinfo` file!

So, I've moved the switch out of the global pipeline default and added it only to the invocation of the linters' run.  😝

webbnh commented:
This PR is now ready for review.

portante reviewed (changes requested):
I am not in favor of this change because it buries it in `build.sh`, which already does too much as it is.

The Alembic checks are really not quite a linter as `black`, etc. are, and the complication another container in a container environment is a complication I'd rather not proliferate (I'd really like to move away from that entirely because I find the environment setup too confusing to debug).  Even reviewing this to understand why host networking is needed for the tox invocation, and why environment variables are added to tox.ini highlight my concerns.

If this was packaged as a "pre-commit" I'd be all for it.

Or if we changed this nature of this check entirely to use the PostgreSQL image as is, mounting our code into a container running that image, where we invoke our check from within that container to avoid this host networking requirement making it a bit more self-contained.

dbutenhof reviewed (approved):
Functionally looks OK, but it does bother me a bit that we're duplicating all the podman options and the PostgreSQL version and tag in various places. I like that this is a fairly simple standalone change, but we should try to address this at some point.

webbnh commented:
> I am not in favor of this change because it buries it in `build.sh`, which already does too much as it is.

The intent (or, at least, my intent...) for `build.sh` is that it should basically "do everything", so that a developer can run "the build" locally in the same way that the CI runs it.  If you buy into that idea, then the Alembic check _should_ be invoked from `build.sh` or from something that it runs.

> The Alembic checks are really not quite a linter as `black`, etc. are

Before starting this work, you and I discussed how we might invoke the Alembic checks.  We agreed that they aren't really unit tests, and we concluded that they didn't fit with the functional tests (since they don't require a running Pbench Server); and so we agreed that we would invoke them from the same part of the build where we run the linters.  I apologize for calling the Alembic check "a linter" -- I'm happy to change the terminology that I'm using, but I think it's appropriate to invoke it after Black, Flake8, etc.

> the complication another container in a container environment is a complication I'd rather not proliferate (I'd really like to move away from that entirely because I find the environment setup too confusing to debug). Even reviewing this to understand why host networking is needed for the tox invocation, and why environment variables are added to tox.ini highlight my concerns.

With the infrastructure provided by the previous PR, invoking a second container from inside a first is now a seamless and transparent operation.  I'm sorry that you find the set up confusing -- I would be happy to alleviate that if you can suggest how.

I believe that host networking is required to allow communication between the invocation of the Alembic commands in the `jenkins/run` container and the database engine in the PostgreSQL container.  (The use of that switch is just a "lift-and-shift" from `jenkins/run-alembic-migrations-check`.)

As for the two environment variables that I added to the Tox environment, `CONTAINER_HOST` is what allows the creation of a second container from inside the first, and `WORKSPACE_TMP` is used by `jenkins/run` to provide the `$HOME` directory inside the container.  (I can see how the name might not be the best, but in the CI environment `WORKSPACE_TMP` is defined by Jenkins for us; otherwise, the human user can define it if s/he cares, and if it's not defined then `jenkins/run` provides ephemeral storage for it...so it really _is_ a "temporary workspace".)

> If this was packaged as a "pre-commit" I'd be all for it.

We don't currently have a server-side (e.g., on GitHub) pre-commit check, right?  So, packaging it as a pre-commit which only some people run doesn't really solve much.  (And, is there some reason why the Alembic check should be handled differently from Black, Flake8, and iSort?...it might be cool to have them all set up as a server-side precommit step and get them off the testing path.)

> if we changed this nature of this check entirely to use the PostgreSQL image as is, mounting our code into a container running that image, where we invoke our check from within that container to avoid this host networking requirement making it a bit more self-contained.

I'm willing to work on that, but as we discussed in today's overflow meeting, it's got some challenges:  we either need to install and run the PostgreSQL server in the container that `jenkins/run` normally uses, or we need to create an environment in the PostgreSQL container in which we can run the Alembic check, and it's not obvious that either of those approaches is simpler than running the Alembic check in one container while PostgreSQL runs in its own container.

> it does bother me a bit that we're duplicating all the podman options and the PostgreSQL version and tag in various places.

Me, too!  But, this PR doesn't increase the number of distinct places where we have that knowledge.

npalaska reviewed (approved):
I am not very familiar with the implication of this change here so I'll leave that to others, but as Dave said its a fairly simple standalone change so I am inclined to say gtg and since you have already worked on it and tested it. If we come across issues where we can not debug the alembic migration test in a functional ci environment then we can revisit this to have it as a `pre-commit` hook or we might think about something else at that point, _I guess_.
